core/src/main/java/com/linecorp/armeria/common/multipart/MultipartFile.java

```diff
@@ -54,7 +57,49 @@ static MultipartFile of(String name, String filename, Path path) {
         requireNonNull(name, "name");
         requireNonNull(filename, "filename");
         requireNonNull(path, "path");
-        return new DefaultMultipartFile(name, filename, path);
+        return of(name, filename, path, HttpHeaders.of());
+    }
+
+    /**
+     * Creates a new {@link MultipartFile}.
+     * @param name the name parameter of the {@code "content-disposition"}
+     * @param filename the filename parameter of the {@code "content-disposition"}
+     *                 header.
+     * @param file the file that stores the {@link BodyPart#content()}.
+     * @param headers HTTP part headers.
+     */
+    static MultipartFile of(String name, String filename, File file, HttpHeaders headers) {
+        requireNonNull(name, "name");
+        requireNonNull(filename, "filename");
+        requireNonNull(file, "file");
+        requireNonNull(headers, "headers");
+        return of(name, filename, file.toPath(), headers);
+    }
+
+    /**
+     * Creates a new {@link MultipartFile}.
+     * @param name the name parameter of the {@code "content-disposition"}
+     * @param filename the filename parameter of the {@code "content-disposition"}
+     *                 header.
+     * @param path the path that stores the {@link BodyPart#content()}.
+     * @param headers HTTP part headers.
+     */
+    static MultipartFile of(String name, String filename, Path path, HttpHeaders headers) {
+        requireNonNull(name, "name");
+        requireNonNull(filename, "filename");
+        requireNonNull(path, "path");
+        requireNonNull(headers, "headers");
+
+        final HttpHeaders newHeaders;
+        if (headers.contentType() == null) {
+            final String guessedContentType = URLConnection.guessContentTypeFromName(filename);
+            final MediaType contentType = guessedContentType != null ?
+                                          MediaType.parse(guessedContentType) : MediaType.OCTET_STREAM;
+            newHeaders = headers.withMutations(builder -> builder.contentType(contentType));
+        } else {
+            newHeaders = headers;
+        }
```

Comment:
Why did you fill the content type from the file name if null? I think `headers.contentType()` should represent the value in headers.

Reply:
> Why did you fill the content type from the file name if null?

I took the approach of [BodyPartBuilder](https://github.com/line/armeria/blob/main/core/src/main/java/com/linecorp/armeria/common/multipart/BodyPartBuilder.java#L155-L157) which seems set content type when it's null.

> I think headers.contentType() should represent the value in headers.

Then is it better to just save the given `headers` even if it is empty? 🤔 
e.x) when called from `of(name, filename, file)` or `of(name, filename, path)`

Reply:
After second thought, I applied your comment but I'm not sure I fully understand it.
PTAL 🙇🏻‍♂️

Reply:
> I took the approach of [BodyPartBuilder](https://github.com/line/armeria/blob/main/core/src/main/java/com/linecorp/armeria/common/multipart/BodyPartBuilder.java?rgh-link-date=2023-08-10T16%3A04%3A58Z#L155-L157) which seems set content type when it's null.

They are different. BodyParBuilder sets the default content type if absent while creating new headers. But your previous code inferred the content type from the file name.

> Then is it better to just save the given headers even if it is empty? 🤔

I think so. It would be better to deliver the  `content-type` header to users as is.

>  applied your comment but I'm not sure I fully understand it.
PTAL 🙇🏻‍♂️

It looks good. :-)

Reply:
Thank you for the detailed guide 😄